Create common system for tile toolbars (PT-185447570)

bgoldowsky opened the pull request:
Implements a new scheme for defining, creating, and placing toolbars for tiles. See PT-185447570.
This scheme is applied to the Text tile in this PR; other tiles can be converted to this system in the future.

Testing notes: 
There should be few user-visible changes.  However, toolbar styling is updated to match more current designs. It should track the position of the tile more closely, without "jumping" when the tile resizes. The toolbar will flip to the top of the tile if there is not enough room for it at the bottom, which is different from previous behavior.  This should be more convenient for tiles that fit within the screen height, but could cause problems if tiles are so tall that neither the top nor the bottom is in view.

scytacki commented on src/components/toolbar/toolbar-button-manager.tsx:
```diff
@@ -0,0 +1,66 @@
+      toolbarButtonInfos.set(tileType, tileButtons);
+    }
+    infos.forEach((info) => {
+      tileButtons.set(info.name, info);
```
Since it would be odd if two buttons with the same name were added, I'd recommend doing a `tileButtons.has(` check first. And if there already is a button with the same name then print a warning with its info (minus the component).

scytacki commented on src/components/toolbar/tile-toolbar-button.tsx:
```diff
@@ -0,0 +1,26 @@
+  Icon: React.FC<React.SVGProps<SVGSVGElement>>;
+  onClick: (e: React.MouseEvent) => void;
+  selected?: boolean; // if undefined, considered to be false
+  enabled?: boolean;  // if undefined, considered to be true!
```
It can cause tricky problems when something is considered true when undefined, so I'd suggest changing this property name to be `disabled`.  Slack post: https://concord-consortium.slack.com/archives/C0303QUNU/p1664938148115039

scytacki reviewed (changes requested):
This looks really good. I found a few things I'd recommend changing. I'm busy today, but I should have a lot of time tomorrow if you want to discuss any of this live.

bgoldowsky commented on src/components/toolbar/toolbar-button-manager.tsx:
```diff
@@ -0,0 +1,66 @@
+  title: string, // user-visible tooltip for the button
+  component: React.ComponentType<IToolbarButtonProps>, // component to render
+  keyHint?: string, // If set, displayed to the user as the hotkey equivalent
+  defaultPosition?: number, // If set, shown on the tile's default toolbar in this position
```
That makes sense, I hadn't considered how to add a `['dataset-view', 'Table']` to the default set.

Alternatively, could there be a global default 'store' so that `useSettingFromStores` would just return that default if there's nothing more specific, and a separate place for defaults is not needed?

bgoldowsky commented:
@scytacki - all comments have been addressed, ready for another look.

scytacki reviewed (approved):
This looks great to me. 👍

tealefristoe commented on src/components/toolbar/toolbar-button-manager.tsx:
```diff
@@ -0,0 +1,70 @@
+//
+// Tiles and plugins can register buttons by specifying their basic attributes
+// and the Component to render for the button.
+// These buttons are rendered as part of the tile so they can make use information provided
```
```suggestion
// These buttons are rendered as part of the tile so they can make use of information provided
```

tealefristoe commented on src/components/toolbar/toolbar-button-manager.tsx:
```diff
@@ -0,0 +1,70 @@
+  return toolbarButtonInfos.get(tileType)?.get(buttonName);
+}
+
+function prettyPrint(info: IToolbarButtonInfo|undefined) {
```
```suggestion
function prettyPrint(info?: IToolbarButtonInfo) {
```

tealefristoe reviewed (approved):
Looks great! I appreciate the many comments that help orient new engineers to the big picture strategy 👍

tejal-shah commented:
@bgoldowsky , I'm seeing issues with the placement of the toolbar, especially when the amount of text in the text tile is more than what's shown and the user uses the up/down arrow to navigate through the lines. Please see these two screencasts to see the issues:
https://screenpal.com/watch/c0QtbwV5Q1P
https://screenpal.com/watch/c0QtbzV5QiL

tejal-shah reviewed (changes requested):
Placement of the toolbar seems to be not right in some cases. (Added links to a couple of screencasts showing the issues)

bgoldowsky commented:
@tejal-shah good catch on the positioning issues.  There are two different things going on:

* The issue seen in your first video is due to a pre-existing bug in the text tile - when placed side by side with a fixed-height tile like the graph, adding lengthy text causes it to go into a weird state. You can see that not only the toolbar, but also the drag handle at the upper right, and the resize handle at the lower right are affected.  I've modified this PR so that the toolbar is not dragged along in this case, but the underlying issue should be added to PT as its own story.

* The second video shows that the current strategy of placing the toolbar at the top or bottom of the tile is not great if the tile is taller than the user's screen.  I am not sure what the desired fix is.  We could limit the height of the text tile, or change the positioning strategy to allow it to be positioned at places other than the top or bottom.  Or we could decide this is an unusual case not worth fixing.  Perhaps @lbondaryk or @mtirenin could chime in on this?

scytacki commented:
For the second bullet, I'd personally vote to have the toolbar stay visible at the bottom inside the view port of the CLUE document if it can't fit at the top or bottom of the tile.  This means it is going to overlap the content of the tile. This is basically the behavior of the toolbar on master currently.

mtirenin commented:
I'd agree with Scott re: toolbar placement.

bgoldowsky commented:
Changed placement strategy to try bottom first, then top, then failing that to place it on top of content at the bottom of the window.

The middle strategy, placing it at the top, is different from master which does not try that option.

Adjusted icon colors to match spec (selected & not selected have different foreground as well as background).

bgoldowsky commented:
@tejal-shah Not sure why the one regression Cypress test failed. It doesn't seem related to these changes, and runs fine locally.

tejal-shah reviewed (approved):
@bgoldowsky , thank you! This looks much better now.
